File: ios/Runner/bluetooth/BluePebbleDevice.swift
//
//  BluePebbleDevice.swift
//  Runner
//
//  Created by crc32 on 17/10/2021.
//

import Foundation
import CoreBluetooth
import CobbleLE
class BluePebbleDevice {
    
    let peripheral: CBPeripheral
    var leMeta: GAPMeta?
    
    init(peripheral: CBPeripheral, advertiseData: [UInt8]?) {
        self.peripheral = peripheral
        if advertiseData != nil {
            leMeta = GAPMeta(data: advertiseData!)
        }
    }
    
    func toPigeon() -> PebbleScanDevicePigeon {
        let pigeon = PebbleScanDevicePigeon()
        pigeon.name = peripheral.name
        pigeon.address = NSNumber(value: peripheral.identifier.uuidString.hashValue)
        
        if leMeta?.major != nil {
            pigeon.version = "\(leMeta!.major!).\(leMeta!.minor!).\(leMeta!.patch!)"
        }
        
        pigeon.serialNumber = leMeta?.serialNumber
        
        if leMeta?.color != nil {
            pigeon.color = NSNumber(value: leMeta!.color!)
        }
        
        if leMeta?.runningPRF != nil {
            pigeon.runningPRF = NSNumber(value: leMeta!.runningPRF!)
        }
        
        if leMeta?.firstUse != nil {
            pigeon.firstUse = NSNumber(value: leMeta!.firstUse!)
        }
        return pigeon
    }
    
    func toStoredPebbleDevice() -> StoredPebbleDevice {
        return StoredPebbleDevice(name: self.peripheral.name!, identifier: self.peripheral.identifier, serialNumber: self.leMeta?.serialNumber, color: self.leMeta?.color)
    }
}

File: android/app/src/main/kotlin/io/rebble/cobble/bluetooth/BluePebbleDevice.kt
package io.rebble.cobble.bluetooth

import android.bluetooth.BluetoothDevice
import android.bluetooth.le.ScanResult
import android.os.Build
import androidx.annotation.RequiresApi
import io.rebble.cobble.pigeons.Pigeons
import io.rebble.cobble.util.macAddressToLong

@OptIn(ExperimentalUnsignedTypes::class)
class BluePebbleDevice {
    val bluetoothDevice: BluetoothDevice
    val leMeta: LEMeta?

    constructor(device: BluetoothDevice) {
        bluetoothDevice = device
        leMeta = null
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(scanResult: ScanResult) {
        bluetoothDevice = scanResult.device
        leMeta = scanResult.scanRecord?.bytes?.let { LEMeta(it) }
    }

    constructor(device: BluetoothDevice, scanRecord: ByteArray) {
        bluetoothDevice = device
        leMeta = LEMeta(scanRecord)
    }

    fun toPigeon(): Pigeons.PebbleScanDevicePigeon {
        return Pigeons.PebbleScanDevicePigeon().also {
            it.name = bluetoothDevice.name
            it.address = bluetoothDevice.address

            if (leMeta?.major != null) {
                it.version = "${leMeta.major}.${leMeta.minor}.${leMeta.patch}"
            }
            if (leMeta?.serialNumber != null) {
                it.serialNumber = leMeta.serialNumber
            }
            if (leMeta?.color != null) {
                it.color = leMeta.color.toLong()
            }
            if (leMeta?.runningPRF != null) {
                it.runningPRF = leMeta.runningPRF
            }
            if (leMeta?.firstUse != null) {
                it.firstUse = leMeta.firstUse
            }
        }
    }

    override fun toString(): String {
        var result = "<${this::class.java.name} "
        for (prop in this::class.java.declaredFields) {
            result += "${prop.name} = ${prop.get(this)} "
        }
        result += ">"
        return result
    }
}